Enh/analysis v3 upgrades

stephlazar opened the pull request:
Mainly upgrades to analysis_v3 which is needed to analyze RB.
Also upgrades the single and two-qubit randomized benchmarking measure functions.
Fixes bug in analysis_toolbox.get_timestamps_in_range which failed when label was an iterable.
Improves the RamseyAddPulseAnalysis in analysis_v2 to save the analysis results from both ramsey fits.

Commit:
Removes get_hdf_param_value from PipelineDataAnalysis.
Rewrites get_data_from_timestamp_list to use get_params_from_hdf_file from helper_functions.py.
Changes input parameters cal_points_list to cal_points in rotate_1d_array and rotate_iq.
Makes do_plotting default True in RabiAnalysis and SingleQubitRBAnalysis.
Makes get_hdf_param_value try to first evaluate the parameter.
Fixes ylabel = keyi in prepare_raw_data_plot_dicts .

Commit:
Makes all nodes specific to a (set of) channels (ex: rotate_iq expects 2 channels only, not list of tuples as before).
Renames average to average_data.
Adds threshold_data node.
Adds default keys_out to all nodes.
Removes dependency on sweep_points in extract_data.

Commit:
Adds readme.pdf to analysis_v3 folder.

Commit:
Adds the helper function add_param and uses it wherever new entries are added to data_dict.
Removes the 'measured_data' nested dictionary in data_dict, and puts raw data dict directly in data_dict.
Places everything else except measurementstring, timestamp, and folder in data_dict['exp_metadata'].

Commit:
Fixes and improvements that make analysis_v3 more user friendly.
Some changes:
 - major improvements to ProcessingPipeline: each node saves keys_out as a nested dictionary,
    can now easily specify keys_out as 'raw' or 'previous node'
 - get_cp_sp_spmap_vnmap_measobjn not also extracts meas_obj_value_names_map
 - get_param can now be passes a path in dictionary
 - code to add nested keys_out now added to add_param
 - adds flatten_list_func and get_sublst_with_all_strings_of_list to helper_functions.py
 - fixes bugs in threshold_data

Commit:
Enhances ProcessingPipeline to specify unique keys_out or container_keys_out throughout the
   pipeline if update_key == False, else overwrite data.
Adds PipelineDataAnalysis_multi_timestamp and makes PipelineDataAnalysis single file only.
   (the former currently only concatenates raw data).

Removes num_bins input parameter from average_data and get_std_deviation, and replaces it with
   shape and averaging_axis.
Fixes a bug in the classes in data_processing.py where all plots in data_dict['plot_dict'] were
   being plotted at each call to plot.
Adds option to not produce raw data plots in RabiAnalysis and SingleQubitRBAnalysis.

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
Updates analysis_v3/readme.pdf.

Commit:
Removes dependence on cal_points. Framework now works also if cal_points is None.
Renames get_cp_sp_spmap_vnmap_measobjn to get_measobj_properties, where the user can now choose
   what variable to extract.
Adds prepare_2d_raw_data_plot_dicts and prepare_2d_plot_dicts.
Renames prepare_raw_data_plot_dicts to prepare_1d_raw_data_plot_dicts.
Improved SingleQubitRBAnalysis plotting; adds error bars.
Adds option to not reshape data in average_data and get_std_deviation.

Commit:
Adds ResidZZFunc to fitting_models.py.
Adds option to do simultaneous fits to analysis_v3.
Adds prepare_residzzfit_dict and prepare_rbleakagefit_dict to fitting.py.
Adds inverted_threshold falg to Multiplexed_Readout_Analysis.
Adds '_' to latest_data when return_timestamp == True.
Renames fig_name to figure_name in analysis_v3.
Add separate fit and plot for the f-state probability in SingleQubitRBAnalysis.
Adds option to append_key or update_key to add_param.
Uses add_param in get_params_from_hdf_file.
Adds prepare_fit_plot_dicts to plotting.py.
Adds option to pass in keys_out_container to add_node in ProcessingPipeline.
Adds add_transform_data_node to ProcessingPipeline.

Commit:
Improves get_timestamps_in_range to correctly find files that match a list of labels.
Adds call to save_processed_data to Singleshot_Readout_Analysis_Qutrit.
Adds option to save_processed_data with hdf_group_name_suffix in RamseyAnalysis.
Improves residual zz plot in RamseyAddPulseAnalysis.
Adds correct_readout to data_processing.py and ProcessingPipeline.
Improves prepare_fit_plot_dicts to check each fit_dict for the keys legend_label and params_to_print.
Adds prepare_residzz_fit_dict, prepare_expdamposc_fit_dict, prepare_joint_residzz_fit_dict to fitting.py.
Adds descriptive comments to how the correlation is done UHFQC_classifier_detector.

Commit:
Fixes bug in SingleQubitRBAnalysis; not d parameter was used, it was always analyzing for single qubits.
Fixes bug in MultiQubit_SingleShot_Analysis.

Commit:
Adds assumptions to the docstring of correct_readout.

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
Adds docstring to SweepPoints.

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

# Conflicts:
#	pycqed/analysis/analysis_toolbox.py
#	pycqed/analysis_v2/timedomain_analysis.py
#	pycqed/measurement/detector_functions.py

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
Removes bug in SingleQubitRBAnalysis.prepare_fitting where for do_simple_fit, 'scale_covar'
   should not be set to False.
Adds get_channel_names_from_timestamp and get_sweep_points_from_timestamp.
Adds data_files to the data_dict in get_params_from_hdf_file.
Updates data_dict with the input parameter options_dict in extract_data.

Commit:
Fixes bug in helper_functions.py/add_param which could not append floats to to existing
   entry in data dict.

Commit:
Fixes pipeline creation and metadata in measure_two_qubit_randomized_benchmarking.
Fixes bug in helper_functions.py/get_sublst_with_all_strings_of_list where the
   original order was not preserved.

Commit:
Fixes pipeline creation and metadata in QuDev_transmon.measure_randomized_benchmarking.

Commit:
Merge branch 'qudev_master' into enh/analysis_v3_upgrades

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
Merge branch 'enh/analysis_v3_upgrades' into devel/analysis_v3

Commit:
Fixes bug in QuDev_transmon.measure_randomized_benchmarking and
   multi_qubit_module.py.measure_two_qubit_randomized_benchmarking where the node
   average_data was created with the wrong name.

antsr commented on pycqed/measurement/multi_qubit_module.py:
```diff
@@ -304,22 +301,32 @@ def get_multi_qubit_prep_params(prep_params_list):
-                      vn in qb.int_avg_classif_det.value_names]
-            for qb in qubits}
-    elif det_type == 'inp_avg_det':
+            qb.name: hlp_mod.get_sublst_with_all_strings_of_list(
+                multi_uhf_det_func.value_names,
+                qb.int_avg_classif_det.value_names)
+            for qb in mobjs}
+    elif multi_uhf_det_func.detectors[0].name == 'UHFQC_input_average_detector':
         meas_obj_value_names_map = {
-            qb.name: [vn + ' ' + qb.instr_uhf() for
-                      vn in qb.inp_avg_det.value_names]
-            for qb in qubits}
+            qb.name: hlp_mod.get_sublst_with_all_strings_of_list(
+                multi_uhf_det_func.value_names, qb.inp_avg_det.value_names)
+            for qb in mobjs}
     else:
         meas_obj_value_names_map = {
-            qb.name: [vn + ' ' + qb.instr_uhf() for
-                      vn in qb.int_avg_det.value_names]
-            for qb in qubits}
+            qb.name: hlp_mod.get_sublst_with_all_strings_of_list(
+                multi_uhf_det_func.value_names, qb.int_avg_det.value_names)
+            for qb in mobjs}
+
+    meas_obj_value_names_map.update({
+        name: [name] for name in
+        [vn for vn in multi_uhf_det_func.value_names if vn not in
+         hlp_mod.flatten_list(list(meas_obj_value_names_map.values()))]})
+
```
I can't judge if this will break any existing analysis.
What have you tested this on?

stephlazar replied:
@antsr  "meas_obj_value_names_map" and hence this method is only used for analysis_v3. I wrote this helper function for better integration of measurements into analysis_v3. This helper function is currently only used in measure_two_qubit_randomized_bechmarking.

stephlazar replied:
Also the rb measurements were tested on BF1 together with @MichKe95 .

antsr commented on pycqed/measurement/sweep_points.py:
```diff
@@ -4,30 +4,77 @@
+        sp.add_sweep_parameter(f'lengths_{qb}', np.arange(100),
+        'V', 'Pulse length')
+    """
+    def __init__(self, param_name=None, values=None, unit='', label=None):
```
did you check that rename `param_label` to `label` did not break any existing code

stephlazar replied:
Yes. It is ok.

antsr commented on pycqed/measurement/sweep_points.py:
```diff
@@ -4,30 +4,77 @@
+    for qb in ['qb1', 'qb2', 'qb3']:
+        sp.add_sweep_parameter(f'lengths_{qb}', np.arange(100),
+        'V', 'Pulse length')
+    """
```
Useful docstring!

However an example that sweeps an amplitude to 99V and pulse length to 99 V! might be a bit confusing.

stephlazar replied:
Thanks :)

antsr reviewed (changes requested):
I did not check the analysis v3 changes in detail, but the other changes seemed mostly reasonable to me.
There are a few minor comments that you could address before merging.

Commit:
Merge branch 'qudev_master' into enh/analysis_v3_upgrades

Commit:
Fixes docstring in SweepPoints where the example for a pulse_length sweep had units of "V" instead of "s".

Commit:
Improves the docstring of SweepPoints:
 - gives realistic sweep values in the example
 - makes the pulse length sweep the inner sweep dimension to reflect what we typically do
